org.uqbar.project.wollok.typeSystem/src/org/uqbar/project/wollok/typesystem/constraints/ConstraintBasedTypeSystem.xtend

```diff
@@ -176,23 +184,45 @@ class ConstraintBasedTypeSystem implements TypeSystem, TypeProvider {
 	}
 
 	def classType(WClass clazz) {
-		new ClassBasedWollokType(clazz, this)
+		genericTypes.get(clazz.fqn) ?: new ClassBasedWollokType(clazz, this)
+	}
+
+	def genericType(WClass clazz, String... typeParameterNames) {
+		genericTypes.get(clazz.fqn) ?: (
+			new GenericType(clazz, this, typeParameterNames) => [
+				genericTypes.put(clazz.fqn, it)
+			]
+		)
 	}
 
 	override objectType(EObject context, String objectFQN) {
 		finder.getCachedObject(context, objectFQN).objectType
 	}
 
+	/**
+	 * Before constructing a class type, check if the provided FQN is known to be a generic type.
+	 * If so, return the known generic type.
+	 * Otherwise create a simple class type. 
+	 */
 	override classType(EObject context, String classFQN) {
-		finder.getCachedClass(context, classFQN).classType
+		genericTypes.get(classFQN) ?: finder.getCachedClass(context, classFQN).classType
+	}
+
+	/**
+	 * Build a generic type and save it, so that we know which concrete types are known to be generic.
+	 */
+	override genericType(EObject context, String classFQN, String... typeParameterNames) {
+		finder.getCachedClass(context, classFQN).genericType(typeParameterNames) => [
+			genericTypes.put(classFQN, it)
+		]		
 	}
 
 	def getAllTypes() {
 		if (allTypes === null) {
 			// Initialize with core classes and wkos, then type system will add own classes incrementally.
 			allTypes = newHashSet
-			allTypes.addAll(allCoreClasses.map[new ClassBasedWollokType(it, this)])
```

Comment:
Avoid manual creation of ClassBasedWollokType, as you do not know if it is generic.